rust/src/adt.rs

```diff
@@ -0,0 +1,909 @@
+// SPDX-License-Identifier: MIT
+use core::ffi::*;
+use core::mem::size_of;
+
+use crate::c_size_t;
+
+#[derive(Debug)]
+pub enum AdtError {
+    NotFound = -1,
+    BadOffset = -4,
+    BadPath = -5,
+    BadNCells = -14,
+    BadValue = -15,
+    BadLength = -20,
+}
+
+const ADT_ALIGN: usize = 4;
+const ADT_MAX_DEPTH: usize = 8;
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct ADTNode {
+    property_count: u32,
+    child_count: u32,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct ADTProperty {
+    name: [c_char; 32],
+    size: u32,
+    value: [u8],
+}
+
+#[repr(C, packed(1))]
+pub struct ADTSegmentRanges {
+    phys: u64,
+    iova: u64,
+    remap: u64,
+    size: u32,
+    unk: u32,
+}
+
+/// Check that our address is properly aligned
+fn check_ptr(ptr: usize) -> Result<(), AdtError> {
+    if ptr % ADT_ALIGN != 0 {
+        Err(AdtError::BadOffset)
+    } else {
+        Ok(())
+    }
+}
+
+/// Determine if an ADT node's name is equal to some string bounded by a length.
+/// This is required to match the semantics expected when doing a full ADT path
+/// trace.
+fn node_names_equal(a: &str, b: &str, len: usize) -> bool {
+    if a == &b[..len] {
+        return true;
+    }
+
+    if b[..len].contains('@') && a.as_bytes().get(len) == Some(&b'@') {
+        return true;
+    }
+
+    false
+}
+
+fn get_cells(src: &[u32]) -> u64 {
+    let mut val: u64 = 0;
+
+    src.iter()
+        .enumerate()
+        .for_each(|(i, &x)| val |= (x as u64) << (32 * i));
+
+    val
+}
+
+/// Retrieve the register container (addr, size) of a given node. Each addr
+/// and size is a contiguous array of u32s with `#address-cells` or `#size-cells`
+/// elements. The entire `reg` property is similarly a contiguous array of these
+/// containers [(addr, size), (addr, size), ...]. The `i` argument is an index
+/// into this array.
+pub fn get_reg_container(
+    nodes: &[Option<&'static ADTNode>],
+    pname: &str,
+    i: i32,
+) -> Result<(u64, u64), AdtError> {
+    let head = unsafe { ADTNode::from_ptr(adt as *const ADTNode)? };
+
+    if nodes[0].is_some() {
+        if nodes[0].unwrap().as_ptr() == unsafe { adt as *const u8 } {
```

Comment:
if let Some(...) = nodes[0]

Reply:
Added sanity check made it safe to just unwrap nodes[0] unconditionally